### ISSUE TITLE
CMake can not find 'Threads' (pthreads) when targeting Webassembly

### ISSUE THREAD

**Issue by PetterS** [kripken/emscripten#5062]
CMake can not find 'Threads' (pthreads) when targeting Webassembly. This seems to be true for both master and incoming.

Minimum example:

CMakeLists.txt

	cmake_minimum_required(VERSION 3.0)

	set(EMSCRIPTEN_FEATURES "-s WASM=1 -s BINARYEN_METHOD=native-wasm -s BINARYEN_TRAP_MODE='allow'")
	set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${EMSCRIPTEN_FEATURES}")
	set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${EMSCRIPTEN_FEATURES}")
	set(CMAKE_LINKER_FLAGS "${CMAKE_LINKER_FLAGS} ${EMSCRIPTEN_FEATURES}")

	find_package(Threads)
	if (Threads_FOUND)
		message("-- OK.")
	else ()
		message(FATAL_ERROR "Threads not found.")
	endif ()

	add_executable(main main.cpp)

main.cpp

	int main(){}

The command line I used:

	rm -rf * && cmake -DCMAKE_TOOLCHAIN_FILE=~/emsdk_portable/emscripten/master/cmake/Modules/Platform/Emscripten.cmake ~/Dropbox/Home/emtest && make

Obviously change the source path and n.b. the first rm command that I use to clear everything.

This gives the error "Threads not found". If I comment out the EMSCRIPTEN_FEATURES line in CMakeLists.txt, it compiles fine.

**Comment by PetterS** [comment]
CMakeError.log reveals

	ERROR:root:We cannot build the native system library in /home/petter/.emscripten_ports/binaryen/binaryen-version_30 w
	hen under the influence of emmake/emconfigure. To avoid this, create system dirs beforehand, so they are not auto-built
	 on demand. For example, for binaryen, do python embuilder.py build binaryen

**Comment by kripken** [comment]
Not sure about that error, but more generally, wasm doesn't have threads yet.

Does `python embuilder.py build binaryen` work as it suggests to try?

**Comment by juj** [comment]
For threading support, just add the linker flag -s USE_PTHREADS=1 when
targeting asm.js. The CMake threads package search is not currently used,
so you can skip over those checks in CMake.

2017-03-21 1:19 GMT+02:00 Alon Zakai <notifications@github.com>:

**Comment by PetterS** [comment]
juj, thanks, but the thread package search is used by a lot of preexisting library code. It would be nice to see it work.

Kripken, haven't tried that command yet. I suppose I wasn't sure where to run it. I'll have to look into that later.

**Comment by juj** [comment]
If you want to try to get it to work, you can try adding a no-op fake search .cmake file to https://github.com/kripken/emscripten/tree/incoming/cmake/Modules that always succeeds. See the existing files there for an example.

**Comment by yukels** [comment]
juj, you mentioned that pthread support for wasm is depended on WASM 1.1. Do you have any estimation when pthreads will be available for wasm?

**Comment by juj** [comment]
The rough timeline is around summer for development versions in Nightly and later this year for release.

**Comment by PetterS** [comment]
I solved my problem. If I comment out the `message(FATAL_ERROR` line and actually build the project with `WASM=1`, it will build a lot of Binaryen stuff the first time. After that, `find_package(Threads)` works.

**Comment by Lachrista** [comment]
@juj why "-s USE_PTHREADS=1 -s WASM=1"  error “assert not shared.Settings.USE_PTHREADS, 'WebAssembly does not support pthreads'
AssertionError: WebAssembly does not support pthreads” ？
Does WebAssembly have a way to support pthread?

**Comment by dschuff** [comment]
The 'incoming' branches of emscripten have support for threads with wasm, but browsers do not yet support that by default. In order to use them you need to enable SharedArrayBuffer and wasm threads in your browser (I believe Chrome and Firefox support this currently).

**Comment by Lachrista** [comment]
@dschuff Can you tell me which branch it is? Does the original emscripten need to be reinstalled?

**Comment by juj** [comment]
@Lachrista to get to `incoming` branch, you can try with

```
git clone https://github.com/juj/emsdk.git
cd emsdk
emsdk install sdk-incoming-64bit binaryen-master-64bit
emsdk activate sdk-incoming-64bit binaryen-master-64bit
source ./emsdk_env.sh
```

that will get you the `incoming` branch that has support for `-s USE_PTHREADS=1 -s WASM=1` simultaneously. Although the latest precompiled tag should work as well, so 


```
git clone https://github.com/juj/emsdk.git
cd emsdk
emsdk install latest
emsdk activate latest
source ./emsdk_env.sh
```

should be fine and enable wasm+multithreading as well.

**Comment by Lachrista** [comment]
@juj I'll give it a try. Thank you very much.

**Comment by Lachrista** [comment]
@juj emsdk_set_env.sh can't find，so source ./emsdk_env.sh error，What causes this?

**Comment by Lachrista** [comment]
source ./emsdk_env.sh 
Traceback (most recent call last):
  File "./emsdk", line 2337, in <module>
    sys.exit(main())
  File "./emsdk", line 2229, in main
    silentremove(EMSDK_SET_ENV) # Clean up old temp file up front, in case of failure later before we get to write out the new one.
  File "./emsdk", line 1969, in silentremove
    os.remove(filename)
OSError: [Errno 13] Permission denied: 'emsdk_set_env.sh'

**Comment by Lachrista** [comment]
@juj According to the above method,installation sdk-incoming-32bit binaryen-master-32bit，
 it is still unable to compile -s USE_PTHREADS=1 -s WASM=1.
and prompt
“assert not shared.Settings.USE_PTHREADS, 'WebAssembly does not support pthreads'
AssertionError: WebAssembly does not support pthreads” ？
Does WebAssembly have a way to support pthread?“
Can only 64 bits be installed?

**Comment by juj** [comment]
Hmm, I think it should not make a difference whether the installation was 32-bit or 64-bit. After installation of `incoming`, `emcc -v` should give

```
C:\code\emsdk\emscripten\incoming>emcc -v
emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 1.37.33
clang version 5.0.0 (https://github.com/kripken/emscripten-fastcomp-clang/ b84bc62bab5a245ef5d8541bc4b6283b7e0a2347) (https://github.com/kripken/emscripten-fastcomp/ a7339d44260fb6bba38eec2344472ba8d19a07e4) (emscripten 1.37.33 : 1.37.33)
Target: x86_64-pc-windows-msvc
Thread model: posix
InstalledDir: C:\code\emsdk\clang\fastcomp\build_incoming_vs2015_64\RelWithDebInfo\bin
INFO:root:(Emscripten: Running sanity checks)
```

and if you search for `incoming` branch, there no longer does exist a assert not shared.Settings.USE_PTHREADS` line like you refer to:

```
Searching 7467 files for "assert not shared.Settings.USE_PTHREADS"

c:\code\emsdk\emscripten\incoming\emcc.py:
  995:         assert not shared.Settings.USE_PTHREADS, 'no SPLIT_MEMORY with pthreads'

1 match in 1 file
```

Testing out with `incoming` on the following should succeed:

```
C:\code\emsdk\emscripten\incoming>emcc tests\pthread\hello_thread.c -o hello_thread.html -s USE_PTHREADS=1 -s WASM=1
```

Perhaps your installation of `incoming` did somehow not get the latest `incoming`? Try running `git log -n1` in the Emscripten directory to check, which should show something like

```
C:\code\emsdk\emscripten\incoming>git log -n1
commit 5ae0d9fc2b2801605279434d15cba54303ae98a9 (HEAD -> incoming, origin/incoming)
Merge: c2024b1d6 79c74ba90
Author: juj <jujjyl@gmail.com>
Date:   Thu Feb 8 19:41:17 2018 +0200

    Merge pull request #6200 from juj/webgl2_fix_duplicate_gldrawelements

    Webgl2 fix duplicate gldrawelements
```

**Comment by Lachrista** [comment]
@juj 
Well, it has been installed successfully, and you can compile the generated xx.js / xx.js.men / pthread-main.js / xx.wasm,Thank you very much,But I have a problem.
chrome://flags/ ,I've already set up 
(1)shared-array-buffer
(2)All about webassembly
but, when time runs in Google and firefox, it will be wrong."Uncaught TypeError: [object Uint32Array] is not an integer shared typed array.  at Object.initMainThreadBlock ",
I'm very confused. Can you answer that?

**Comment by Lachrista** [comment]
@juj 
The fact is that you can compile it, but you can't run it on your browser.
#include <stdio.h>
int main(){
	printf("hello\n");
}
emcc test.c -s WASM=1 -o test.html
test.html can run normally.
but
emcc test.c -s WASM=1 -s USE_PTHREADS=1 -o test.html
test.html unable to run properly.

error:
TypeError: invalid array type for the operation  test.js:1824:9
at initMainThreadBlock
Can you answer for me?

**Comment by juj** [comment]
Trying out the `test.c` file above, I noticed a regression in `incoming`, fix is in PR #6268. This regression causes the print to not appear in the above code.

Although I don't get the same error that you are getting, either before or after the above PR has been applied.

After the above PR #6268 is applied, the code properly prints hello when building with threading enabled. (testing in Firefox Nightly with pref `javascript.options.shared_memory` set to `true` in `about:config` page)

**Comment by Lachrista** [comment]
@juj 
Can you print out hello from the test.html that you compiled?
But I can't.
emcc test.c -s WASM=1 -s USE_PTHREADS=1 -o test.html
I got four files altogether.
1)pthread-main.js
2)test.html
3)test.html.mem
4)test.js
5)test.wasm
and testing in Firefox Nightly with pref javascript.options.shared_memory set to true in about:config page
but
always error：
TypeError: invalid array type for the operation

I'm very distressed, maybe you can provide an example of -s WASM= 1-s USE_PTHREADS=1. Can I refer to it?
